exonum/src/api/ext.rs

```diff
@@ -0,0 +1,751 @@
+    #[fail(display = "Unknown endpoint ID: {}", _0)]
+    UnknownId(String),
+
+    // TODO: split `serde::Error` / others?
+    /// A request is malformed or otherwise cannot be processed.
+    #[fail(display = "Bad request: {}", _0)]
+    BadRequest(Box<::std::error::Error + Send + Sync>),
+
+    /// Requested resource is not found.
+    #[fail(display = "Not found")]
+    NotFound,
+
+    /// Transaction verification has failed.
+    #[fail(display = "Transaction not verified: {:?}", _0)]
+    VerificationFail(Box<Transaction>),
+
+    /// A transaction processed by API has failed to be sent.
+    #[fail(display = "Failed to send transaction: {}", _0)]
+    TransactionNotSent(
+        #[cause]
+        io::Error
+    ),
+
+    /// Generic server-side error.
+    #[fail(display = "Internal server error: {}", _0)]
+    InternalError(Box<::std::error::Error + Send + Sync>),
+}
+
+impl From<SendError> for ApiError {
+    fn from(err: SendError) -> ApiError {
+        use self::SendError::*;
+
+        match err {
+            VerificationFail(tx) => ApiError::VerificationFail(tx),
+            Io(e) => ApiError::TransactionNotSent(e),
+        }
+    }
+}
+
+/// Alias for the result type used within the `api` module.
+pub type ApiResult<T> = Result<T, ApiError>;
+
+/// Specification of an endpoint.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct Spec {
+    /// Endpoint identifier.
+    pub id: &'static str,
+
+    /// Visibility level of the endpoint.
+    ///
+    /// Endpoints with lesser visibility may be hidden by the transport adapters;
+    /// e.g., an HTTP adapter may serve them on a different port behind a firewall.
+    pub visibility: Visibility,
+}
+
+/// Possible visibility levels of service endpoints.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum Visibility {
+    /// Endpoint should be available to the general audience.
+    Public,
+    /// The access to the endpoint should be restricted.
+    Private,
+}
+
+/// Context supplied to endpoints.
+#[derive(Debug)]
+pub struct Context<'a> {
+    blockchain: &'a Blockchain,
+    mutable: Option<MutContext<'a>>,
+}
+
+impl<'a> Context<'a> {
+    fn new(blockchain: &'a Blockchain) -> Self {
+        Context {
+            blockchain,
+            mutable: None,
+        }
+    }
+
+    fn mutable(blockchain: &'a Blockchain) -> Self {
+        Context {
+            blockchain,
+            mutable: Some(MutContext::new(blockchain)),
+        }
+    }
+
+    /// Gets a snapshot of the current blockchain state.
+    pub fn snapshot(&self) -> Box<Snapshot> {
+        self.blockchain.snapshot()
+    }
+
+    /// Gets the public key used for signing operations by services.
+    pub fn service_public_key(&self) -> &PublicKey {
+        self.blockchain.service_public_key()
+    }
+
+    /// Accesses the mutable context.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if invoked in a constant endpoint.
+    pub fn as_mut(&self) -> &MutContext<'a> {
```

Comment:
So can we avoid to write code that can panic?
Maybe we can split `TypedEndpoint` to the `TypedConstEndpoint` and `TypedReadEndpoint` with the `&Context` and `&mut Context` arguments?